compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/LimitScatterBounds.cpp

```diff
@@ -0,0 +1,281 @@
+      ValueRange{bounded, inbounds},
+      ArrayRef<AffineMap>{boundedMap, boundsMap, boundedMap, boundedMap},
+      indicesIter, [](OpBuilder &bb, Location loc, ValueRange args) {
+        ImplicitLocOpBuilder b(loc, bb);
+        Value zero =
+            b.create<arith::ConstantOp>(b.getZeroAttr(args[0].getType()));
+        Value lower =
+            b.create<arith::CmpIOp>(arith::CmpIPredicate::sge, args[0], zero);
+        Value upper = b.create<arith::CmpIOp>(arith::CmpIPredicate::sle,
+                                              args[0], args[1]);
+        Value within = b.create<arith::AndIOp>(lower, upper);
+        Value sel = b.create<arith::SelectOp>(within, args[0], zero);
+        b.create<linalg::YieldOp>(ValueRange{sel, within});
+      });
+
+  bounded = applyLimits.getResult(0);
+  inbounds = applyLimits.getResult(1);
+
+  // If the index is not implicit we need to reduce away the index vector dim:
+  if (indexVectorDim < indicesTy.getRank()) {
+    ShapedType inTy = cast<ShapedType>(inbounds.getType());
+
+    llvm::SmallVector<AffineExpr> inExprs;
+    llvm::SmallVector<AffineExpr> outExprs;
+    llvm::SmallVector<int64_t> outShape;
+    for (int i = 0; i < inTy.getRank(); ++i) {
+      if (i < indexVectorDim) {
+        inExprs.push_back(b.getAffineDimExpr(outShape.size()));
+        outExprs.push_back(b.getAffineDimExpr(outShape.size()));
+        outShape.push_back(inTy.getDimSize(i));
+      } else {
+        // We need to do a reduction across the last dimension:
+        inExprs.push_back(b.getAffineDimExpr(inTy.getRank() - 1));
+      }
+    }
+
+    AffineMap inMap =
+        AffineMap::get(inExprs.size(), 0, inExprs, b.getContext());
+    AffineMap outMap =
+        AffineMap::get(inExprs.size(), 0, outExprs, b.getContext());
+    Value trueval = b.create<arith::ConstantIntOp>(1, b.getI1Type());
+    Value empty = b.create<tensor::SplatOp>(inTy.clone(outShape), trueval);
+
+    SmallVector<utils::IteratorType> reduceIter(outExprs.size(),
+                                                utils::IteratorType::parallel);
+    reduceIter.push_back(utils::IteratorType::reduction);
+
+    inbounds =
+        b.create<linalg::GenericOp>(
+             TypeRange{empty.getType()}, ValueRange{inbounds},
+             ValueRange{empty}, ArrayRef<AffineMap>{inMap, outMap}, reduceIter,
+             [](OpBuilder &bb, Location loc, ValueRange args) {
+               Value andOp = bb.create<arith::AndIOp>(loc, args[0], args[1]);
+               bb.create<linalg::YieldOp>(loc, andOp);
+             })
+            .getResult(0);
+  }
+
+  if (cast<ShapedType>(inbounds.getType()).getRank() < update0Ty.getRank()) {
+    llvm::SmallVector<bool> isBatchDim(update0Ty.getRank(), true);
+    for (auto dim : updateWindowDims) {
+      isBatchDim[dim] = false;
+    }
+
+    llvm::SmallVector<AffineExpr> inExprs;
+    for (int i = 0; i < isBatchDim.size(); i++) {
```

Comment:
Nit: only check .size() per iteration if it may change.

Reply:
Technically we need the index as well but I changed to an `llvm::enumerate` for ease.